### ISSUE TITLE
Outline

### ISSUE THREAD

**Issue by betatim** [wildtreetech/epfl-osip-2017#1]
Collecting ideas for an afternoon workshop on "open science aspects of research code".

Time slot: 2pm - 6.30pm with a 60min break for a talk on IP.

* participants have whole range of skills and background in different disciplines
* not hardcore coders but researchers
* git, github, continuous integration, documentation generation
* jupyter notebooks
* there are lots of tools which ones should I use?
* recommend a cookie-cutter repo to start from to reduce overhead
* Gael's "premature software engineering" list
* pick python for concreteness but make sure to highlight transferable parts

From talking with @mdeff:

> I think that Python / Jupyter notebooks and tools for collaboration and
> publishing are the most important topics. We can stick to that. Short
> sessions is an interesting idea. To maximize engagement, I was thinking
> of having one project for the whole afternoon, e.g. you start with a
> small analysis in a notebook (or some code the participants bring), add
> versioning, go to a script, introduce tests, add generated
> documentation, and finish the day with a github repo.

Learning goals. As a participant I will learn ...
* ... how to setup a "virtual env" for my project and be able to name advantages of doing so for each new project
* ... what a linter does and how to connect it to the Atom editor
* ... how to use jupyter notebooks to perform interactive exploration
* ... how to package my data sanity checks so that they are easy to run
* ... how to automate running my data sanity checks
* ... how to use `git` and GitHub as a "dropbox" for my code
* ... be able to name some of the tradeoffs between just putting my code out there and trying to create a open-source project with external users
* ... to recognise when code should be moved from my notebook to a module

**Comment by betatim** [comment]
How do we balance the need for concrete stuff which pushes us towards python specifics and keeping it general because not everyone uses python?

matplotlib is similar to matlab so we can use that for simple plotting

**Comment by betatim** [comment]
Git novice material from SWC: http://swcarpentry.github.io/git-novice/
Python novice material from SWC: http://swcarpentry.github.io/python-novice-inflammation/ (maybe too technical)

**Comment by betatim** [comment]
Mozilla open leadership training material: https://mozilla.github.io/open-leadership-training-series/ this is mostly about how to take the leap towards creating a open-source project where others join you.

**Comment by mdeff** [comment]
Thanks for the organization @betatim!

Some random notes:
* Time slot: 2pm - 6.30pm with a **1h break** (30min for an IP talk + 30min actual coffee break)
* Gaël Varoquaux will talk in the morning BTW. But he'll leave before the workshop.
* Python, cookie-cutter, Jupyter, git, github are good picks from my point-of-view.
* Continuous integration and documentation generation might be too much in the end.
* Aren't Atom and linter too specific? I otherwise like your "what I will learn list". 👍
* Regarding the statistics below (lots of matlab, little R), matplotlib indeed looks like a good pick.
* I don't think "creating an open-source project where others join you" is the priority of a researcher. At least not an immediate one. A more pressing question I envision is "how do I manage the various issues and questions people will inevitably encounter if I publish my code.". To be concrete, maybe we could talk about github issues, and show some examples of real projects? Or how to minimize user issues in the first place, like a good readme and usage guide / notebook, or testing if we go into that.

**Comment by mdeff** [comment]
Out of the 36 selected participants, we got the following statistics:
* 26 use or develop code for their research
* Programming languages:
    * 15 use matlab
    * 12 use Python
    * 6 use Excel
    * 4 use R
    * 9 use others (among which C++, C#, Javascript, Julia, Ruby, SQL, Perl)
* 4 ever contributed to open-source software development
* Tools
    * 14 use versioning (e.g. git, svn, cvs)
    * 6 use automated testing & continuous integration (e.g. jenkins, travis)
    * 6 use generated documentation (from e.g. markdown or restructured text in the code),
    * 9 use notebooks which integrate text, code and figures (e.g. Jupyter, knitr)
    * 9 use virtualization (containers or virtual machines)
* How do you share code:
    * 11 don't produce code
    * 3 don't share their code
    * 14 share upon request (e.g. by email, wetransfer)
    * 11 use github / gitlab

So we indeed got a large range of skills.

**Comment by betatim** [comment]
> Aren't Atom and linter too specific?

Yes and no. I think compared to many other tools it is super easy to setup and gives you benefits in terms of making "better" code, but how to set it up is super specific to the editor+language combination you use :-/ I assume there is a linter for matlab?

Let's put the idea of "creating an open-source project" on the back burner. Maybe better a "how to get started contributing to an existing project". (if there is time, and maybe I will use that as part of my talk in the morning)

Looks like we have 14 people with some experience on version control -> roughly half -> let's have an exercise on collaborating with yourself and others through GitHub where we pair people up. So far I've not seen a really good way to explain git/GitHub to people who are new to it, so let's try this :)

**Comment by betatim** [comment]
find matlab tools for:
* [ ] unit tests
* [ ] linter
* [ ] what editor do people use for matlab?

**Comment by betatim** [comment]
`mlint`: https://ch.mathworks.com/help/matlab/ref/mlint.html?requestedDomain=www.mathworks.com there are (not actively maintained) Atom plugins

**Comment by mdeff** [comment]
To me "creating an open-source project" is better than "how to get started contributing to an existing project", as I think most of the time scientists want to share the code to reproduce their results. Most will never contribute to a larger open-source effort (this may depend on fields though). And by "open-source project" I see two things:

1. A library (used in the lab and outside) which implements a set of methods developed and routinely used by a lab / group of labs for their analyzes.
2. A one-off analysis, like the code used to produce the particular results of a paper.

I would maybe concentrate on 2, where we have a small codebase that we want to publish, and how to avoid the common pitfalls and maximize the impact of the code. What do you think?

**Comment by mdeff** [comment]
Pairing is a good idea. :) Last time I teach it at EPFL it was pretty easy to get people to use it to version their code alone, i.e. commit and push to archive it somewhere, but collaborating was harder. They were lost once they got merge conflicts, did not know how to properly work with branches, etc.

**Comment by mdeff** [comment]
Regarding matlab:
* I think most people just use the provided IDE (which honestly is quite good, in the vein of Jupyter lab).
* Somebody told me there is now built-in facilities for unit tests. (I don't use matlab anymore myself.)

**Comment by betatim** [comment]
> To me "creating an open-source project" is better than "how to get started contributing to an existing project"

👍  We should invent a new name for this. For me "starting an open-source project" == trying to create a project on the scale of scikit-learn and the like :rofl: (probably not a good idea to do that on your first trip to open-source land)

**Comment by mdeff** [comment]
Haha you're right. 🤣 Maybe it should be "openly publish my research code" ?
